matching/src/Pattern.hs

```diff
@@ -27,36 +27,35 @@ import           Data.Functor.Classes  (Eq1 (..), Show1 (..))
 import           Data.Functor.Foldable (Fix (..), cata)
 import           Data.Maybe            (mapMaybe)
 import           Data.Semigroup        ((<>))
-import           Data.Text             (Text)
+import           Data.Text             (Text, pack)
 import           TextShow              (showt)
 import qualified Data.Yaml.Builder as Y
 import qualified Data.ByteString as B
 
 data Column = Column
               { getMetadata :: !Metadata
               , getTerms    :: ![Fix Pattern]
-              } deriving (Eq)
+              }
 
 instance Show Column where
   showsPrec _ (Column _ ts) =
     showString "Column " . showList ts
 
-newtype Metadata = Metadata [Metadata]
-                 deriving (Show, Eq)
+newtype Metadata = Metadata (Int, String -> [Metadata])
```

Comment:
What does the Int represent?
Also we should probably use a record, and store the function as two fields. Usually functions cannot be shown or compared for equality, etc.

Reply:
I, too, would like to know what the Int is. As soon as we have an MVP I think we should take the time to write a few comments on all the types/functions.

Reply:
Can you please clarify what you mean by storing the function as two fields? I'm not sure I really follow what you're suggesting. Regardless, I didn't want to use a map type for this because a 32 bit integer would require 2^32 map entries. I'm aware that this makes it no longer possible to derive certain instances on these types, but it seems that they are not actually used. If you can suggest a better way for me to deal with exponentially many symbols in a sort, or otherwise clarify what you mean in this comment, I will fix it along with the others when I work on this PR tomorrow.

Reply:
It will be made clear when I change this into a record type tomorrow, but the Int is the number of total constructors for this sort.

Reply:
I actually need to change it to an Integer I think.

Reply:
Actually, I need a description of what Metadata does here at all, so I can start deriving it from Kore.  What are the Ints and Strings here?

Reply:
The metadata is a good place to start because I already have some code in my working copy that generates the patterns and the actions. The int is the number of constructors in the sort which this metadata is represented by, and the function takes a pretty printed SymbolOrAlias and returns the list of metadata for each of the sorts of the arguments to that constructor.

Reply:
Basically it's a lazily computed infinite data structure.

Reply:
Ok. I see the point of the function.